django/river/parsing.py

```diff
@@ -0,0 +1,202 @@
+from typing import Any, Dict, List, Optional, Tuple
+from uuid import uuid4
+
+from pydantic import BaseModel as PydanticBaseModel
+from pydantic import Field
+
+
+class BaseModel(PydanticBaseModel):
+    class Config:
+        extra = "forbid"
+
+
+class Join(BaseModel):
+    columns: Tuple[str, str]
+
+
+class Column(BaseModel):
+    id: str
+    table: str
+    column: str
+    joins: List[Join]
+
+
+class Input(BaseModel):
+    script: str
+    concept_map_id: str
+    concept_map: Optional[dict]
+    static_value: Any
+    column: Optional[str]
+
+
+class Condition(BaseModel):
+    action: str
+    column: str
+    value: str = ""
+    relation: str
+
+
+class InputGroup(BaseModel):
+    id: str
+    merging_script: str
+    inputs: List[Input]
+    conditions: List[Condition]
+
+
+class Comment(BaseModel):
+    pass
+
+
+class Attribute(BaseModel):
+    path: str
+    slice_name: str
+    definition_id: str
+    input_groups: List[InputGroup]
+    comments: List[Comment] = []
+
+
+class Filter(BaseModel):
+    relation: str
+    value: str
+    sql_column: str
+
+
+class Definition(BaseModel):
+    class Config:
+        extra = "allow"
+
+    id: str
+
+
+class Resource(BaseModel):
+    # Not sure we actually care about the usual id.
+    # For now, use a randomly generated id
+    id: str = str(uuid4())
+    label: str
+    primary_key_table: Optional[str] = None
+    primary_key_column: Optional[str] = None
+    primary_key_owner: Optional[str] = None
+    definition_id: str
+    logical_reference: str
+    attributes: List[Attribute] = []
+    filters: List[Filter]
+    definition: Optional[Definition] = None
+
+
+class Owner(BaseModel):
+    id: str
+    name: str
+    schema_: str = Field(alias="schema")
+    columns: List[Column]
+
+
+class Credential(BaseModel):
+    host: str
+    port: int
+    database: str
+    model: str
+    login: str = ""
+    password: str = ""
+    owners: List[Owner]
+
+
+class Source(BaseModel):
+    id: str
+    name: str
+    version: str
+    resources: List[Resource]
+    credential: Credential
+    created_at: Optional[str]
+    updated_at: Optional[str]
+    users: Optional[List[str]]
+
+
+def as_old_mapping(source: Source, resource_id: str):
```

Comment:
If I understand correctly, this function takes a Source and returns the mapping only for a resource? I find that a bit weird but if it's what's the simplest for the rest of the code ok

Reply:
yes we don't need to make the transformation for all resources since we analyze a resource mapping at a time.